src/traces/bar/hover.js

```diff
@@ -61,40 +62,63 @@ function hoverOnBars(pointData, xval, yval, hovermode) {
             return Math.max(thisBarMaxPos(di), di.p + t.bardelta / 2);
         };
 
-    function _positionFn(_minPos, _maxPos) {
+    function hoverPositionFn(_minPos, _maxPos) {
         // add a little to the pseudo-distance for wider bars, so that like scatter,
         // if you are over two overlapping bars, the narrower one wins.
         return Fx.inbox(_minPos - posVal, _maxPos - posVal,
             maxHoverDistance + Math.min(1, Math.abs(_maxPos - _minPos) / pRangeCalc) - 1);
     }
 
+    function spikePositionFn(_minPos, _maxPos) {
+        // add a little to the pseudo-distance for wider bars, so that like scatter,
+        // if you are over two overlapping bars, the narrower one wins.
+        return Fx.inbox(_minPos - posVal, _maxPos - posVal,
+            maxSpikeDistance + Math.min(1, Math.abs(_maxPos - _minPos) / pRangeCalc) - 1);
```

Comment:
If the only difference between `spikePositionFn` and `hoverPositionFn` is which `max*Distance` is used, just make that a 3rd arg to the `_positionFn` calls below.

Same with `sizeFn` vs `thisBarSizeFn`, looks like that could be a higher-order function like `sizeFn(whichMaxDist)(di)`